### PR TITLE
Data api additions

### PR DESCRIPTION
**Pull request by phlptp**
### Summary

If merged this pull request will mostly complete the data API in the C shared library interface

### Proposed changes

- add additional functions in the C shared library data API
- start on the C++98 interface for the data API(incomplete, will be fleshed out in the subsequent release)

### CODE REVIEW

File: src/helics/cpp98/Translator.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+#define HELICS_CPP98_TRANSLATOR_HPP_
+
+#include "DataBuffer.hpp"
+#include "helics/helics.h"
+#include "helicsExceptions.hpp"
+
+#include <string>
+
+namespace helicscpp {
+/** object managing a translatorer in the C++98 interface*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:17]:
```suggestion
/** object managing a translator in the C++98 interface*/
```

File: src/helics/cpp98/Translator.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        translator = translatorer.translator;
+        return *this;
+    }
+
+    /** cast operator to get the underlying object*/
+    operator HelicsTranslator() const { return translator; }
+    /** get the underlying HelicsTranslator object*/
+    HelicsTranslator baseObject() const { return translator; }
+    /** check if the translatorer is valid */
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:37]:
```suggestion
    /** check if the translator is valid */
```

File: src/helics/cpp98/Translator.hpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    bool isValid() const { return (helicsTranslatorIsValid(translator) == HELICS_TRUE); }
+    /** get the name for the translatorer*/
+    const char* getName() const { return helicsTranslatorGetName(translator); }
+    /** set a property on a translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:41]:
```suggestion
    /** set a property on a translator
```

File: src/helics/cpp98/Translator.hpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    @param property the name of the property of the translatorer to change
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:42]:
```suggestion
    @param property the name of the property of the translator to change
```

File: src/helics/cpp98/Translator.hpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    void setInfo(const std::string& info)
+    {
+        helicsTranslatorSetInfo(translator, info.c_str(), HELICS_IGNORE_ERROR);
+    }
+    /** get the value of a tag for the translatorer*/
+    const char* getTag(const std::string& tagname) const
+    {
+        return helicsTranslatorGetTag(translator, tagname.c_str());
+    }
+    /** set the value of a tag for the translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:102]:
```suggestion
    /** set the value of a tag for the translator*/
```

File: src/helics/cpp98/Translator.hpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        helicsTranslatorRemoveTarget(translator, dest.c_str(), hThrowOnError());
+    }
+    /** get the interface information field of the translatorer*/
+    const char* getInfo() const { return helicsTranslatorGetInfo(translator); }
+    /** set the interface information field of the translatorer*/
+    void setInfo(const std::string& info)
+    {
+        helicsTranslatorSetInfo(translator, info.c_str(), HELICS_IGNORE_ERROR);
+    }
+    /** get the value of a tag for the translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:97]:
```suggestion
    /** get the value of a tag for the translator*/
```

File: src/helics/cpp98/Translator.hpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        helicsTranslatorAddPublicationTarget(translator, source.c_str(), hThrowOnError());
+    }
+    /** remove a destination target from a cloning translatorer*/
+    void removeTarget(const std::string& dest)
+    {
+        helicsTranslatorRemoveTarget(translator, dest.c_str(), hThrowOnError());
+    }
+    /** get the interface information field of the translatorer*/
+    const char* getInfo() const { return helicsTranslatorGetInfo(translator); }
+    /** set the interface information field of the translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:92]:
```suggestion
    /** set the interface information field of the translator*/
```

File: src/helics/cpp98/Translator.hpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    void addSourcePublication(const std::string& source)
+    {
+        helicsTranslatorAddPublicationTarget(translator, source.c_str(), hThrowOnError());
+    }
+    /** remove a destination target from a cloning translatorer*/
+    void removeTarget(const std::string& dest)
+    {
+        helicsTranslatorRemoveTarget(translator, dest.c_str(), hThrowOnError());
+    }
+    /** get the interface information field of the translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:90]:
```suggestion
    /** get the interface information field of the translator*/
```

File: src/helics/cpp98/Translator.hpp [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+        helicsTranslatorAddInputTarget(translator, dest.c_str(), hThrowOnError());
+    }
+
+    /** add a source target to a cloning translatorer
+    @details all messages coming from a source are copied to the delivery address(es)*/
+    void addSourcePublication(const std::string& source)
+    {
+        helicsTranslatorAddPublicationTarget(translator, source.c_str(), hThrowOnError());
+    }
+    /** remove a destination target from a cloning translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:85]:
```suggestion
    /** remove a destination target from a cloning translator*/
```

File: src/helics/cpp98/Translator.hpp [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    /** add a destination target to a cloning translatorer
+    @details all messages going to a destination are copied to the delivery address(es)*/
+    void addDestinationInput(const std::string& dest)
+    {
+        helicsTranslatorAddInputTarget(translator, dest.c_str(), hThrowOnError());
+    }
+
+    /** add a source target to a cloning translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:79]:
```suggestion
    /** add a source target to a cloning translator
```

File: src/helics/cpp98/Translator.hpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    @details all messages coming from a source are copied to the delivery address(es)*/
+    void addSourceEndpoint(const std::string& source)
+    {
+        helicsTranslatorAddSourceEndpoint(translator, source.c_str(), hThrowOnError());
+    }
+
+    /** add a destination target to a cloning translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:72]:
```suggestion
    /** add a destination target to a cloning translator
```

File: src/helics/cpp98/Translator.hpp [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    @details all messages going to a destination are copied to the delivery address(es)*/
+    void addDestinationEndpoint(const std::string& dest)
+    {
+        helicsTranslatorAddDestinationEndpoint(translator, dest.c_str(), hThrowOnError());
+    }
+
+    /** add a source target to a cloning translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:65]:
```suggestion
    /** add a source target to a cloning translator
```

File: src/helics/cpp98/Translator.hpp [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    /** set a string property on a translatorer
+   @param property the name of the property of the translatorer to change
+   @param val the numerical value of the property
+   */
+    void setString(const std::string& property, const std::string& val)
+    {
+        helicsTranslatorSetString(translator, property.c_str(), val.c_str(), hThrowOnError());
+    }
+
+    /** add a destination target to a cloning translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:58]:
```suggestion
    /** add a destination target to a cloning translator
```

File: src/helics/cpp98/Translator.hpp [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    /** set a property on a translatorer
+    @param property the name of the property of the translatorer to change
+    @param val the numerical value of the property
+    */
+    void set(const std::string& property, double val)
+    {
+        helicsTranslatorSet(translator, property.c_str(), val, hThrowOnError());
+    }
+    /** set a string property on a translatorer
+   @param property the name of the property of the translatorer to change
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:50]:
```suggestion
   @param property the name of the property of the translator to change
```

File: src/helics/cpp98/Translator.hpp [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    const char* getName() const { return helicsTranslatorGetName(translator); }
+    /** set a property on a translatorer
+    @param property the name of the property of the translatorer to change
+    @param val the numerical value of the property
+    */
+    void set(const std::string& property, double val)
+    {
+        helicsTranslatorSet(translator, property.c_str(), val, hThrowOnError());
+    }
+    /** set a string property on a translatorer
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:49]:
```suggestion
    /** set a string property on a translator
```

File: src/helics/cpp98/Translator.hpp [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+        return *this;
+    }
+
+    /** cast operator to get the underlying object*/
+    operator HelicsTranslator() const { return translator; }
+    /** get the underlying HelicsTranslator object*/
+    HelicsTranslator baseObject() const { return translator; }
+    /** check if the translatorer is valid */
+    bool isValid() const { return (helicsTranslatorIsValid(translator) == HELICS_TRUE); }
+    /** get the name for the translatorer*/
```
Comment (nightlark) [inline src/helics/cpp98/Translator.hpp:39]:
```suggestion
    /** get the name for the translator*/
```

File: src/helics/shared_api_library/helicsData.cpp [hunk 17]
```diff
@@ -234,13 +303,240 @@ double helicsDataBufferToDouble(HelicsDataBuffer data)
...
+    }
+    std::string v;
+    helics::valueExtract(helics::data_view(*ptr), helics::detail::detectType(ptr->data()), v);
+
+    auto length = (std::min)(static_cast<int>(v.size()), maxStringLen);
+    std::memcpy(outputString, v.data(), length);
+    if (actualLength != nullptr) {
+        *actualLength = length;
+    }
+    // add a null terminator if possible
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.cpp:414]: I think ending the string with a null terminator should be consistent regardless of buffer space -- probably safer with a null terminator assuming the data is guaranteed to not have null bytes, but if the helics data buffer being converted can contain any data bytes then we should never add a null terminator.

File: src/helics/shared_api_library/helicsData.h [hunk 18]
```diff
@@ -19,6 +19,9 @@ extern "C" {
 /** create a helics managed data buffer with initial capacity*/
 HELICS_EXPORT HelicsDataBuffer helicsCreateDataBuffer(int32_t initialCapacity);
 
+/** check whether a buffer is Valid*/
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.h:22]:
```suggestion
/** check whether a buffer is valid*/
```

File: src/helics/shared_api_library/helicsData.h [hunk 19]
```diff
@@ -34,20 +37,27 @@ HELICS_EXPORT int32_t helicsDataBufferCapacity(HelicsDataBuffer data);
...
-HELICS_EXPORT int32_t helicsIntToBytes(int64_t value, HelicsDataBuffer data);
+HELICS_EXPORT int32_t helicsIntegerToBytes(int64_t value, HelicsDataBuffer data);
 
 /** convert a double to serialized bytes*/
 HELICS_EXPORT int32_t helicsDoubleToBytes(double value, HelicsDataBuffer data);
 
 /** convert a string to serialized bytes*/
 HELICS_EXPORT int32_t helicsStringToBytes(const char* str, HelicsDataBuffer data);
 
+/** convert a raw string(may contain nulls) to serialized bytes*/
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.h:56]:
```suggestion
/** convert a raw string (may contain nulls) to serialized bytes*/
```

File: src/helics/shared_api_library/helicsData.h [hunk 20]
```diff
@@ -58,20 +68,70 @@ HELICS_EXPORT int32_t helicsTimeToBytes(HelicsTime value, HelicsDataBuffer data)
...
+/** get the length of a string if a string were to be retrieved*/
+HELICS_EXPORT int helicsDataBufferStringSize(HelicsDataBuffer data);
+
+/* convert a data buffer to a string*/
+HELICS_EXPORT void helicsDataBufferToString(HelicsDataBuffer data, char* outputString, int maxStringLen, int* actualLength);
+
+/** convert a data buffer to a time*/
+HELICS_EXPORT HelicsTime helicsDataBufferToTime(HelicsDataBuffer data);
+
+/** convert a data buffer to a time*/
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.h:107]:
```suggestion
/** convert a data buffer to a complex object*/
```

File: src/helics/shared_api_library/helicsData.h [hunk 21]
```diff
@@ -58,20 +68,70 @@ HELICS_EXPORT int32_t helicsTimeToBytes(HelicsTime value, HelicsDataBuffer data)
...
+/** get the number of elements that would be required if a vector were retrieved*/
+HELICS_EXPORT int helicsDataBufferVectorSize(HelicsDataBuffer data);
+
+/** convert a data buffer to double vector values*/
+HELICS_EXPORT void helicsDataBufferToVector(HelicsDataBuffer data, double values[], int maxlen, int* actualSize);
+
+/** convert a data buffer to complex double vector values
+@param data the buffer containing data
+@param values the storage for the converted data
+@param maxlen the number of complex values the the values vector can hold
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.h:122]:
```suggestion
@param maxlen the number of complex values that the values vector can hold
```

File: src/helics/shared_api_library/helicsData.cpp (hunk 17 above)
Comment (phlptp) [reply src/helics/shared_api_library/helicsData.cpp]: I split this into two distinct functions one guarantees a null terminator, the other does not do anything with null terminators.    And the string size method was modified so the size returned includes space for a null terminator.

File: src/helics/shared_api_library/helicsData.h [hunk 22]
```diff
@@ -58,20 +68,76 @@ HELICS_EXPORT int32_t helicsTimeToBytes(HelicsTime value, HelicsDataBuffer data)
...
+HELICS_EXPORT HelicsBool helicsDataBufferToBoolean(HelicsDataBuffer data);
+
+/** convert a data buffer to a char*/
+HELICS_EXPORT char helicsDataBufferToChar(HelicsDataBuffer data);
+
+/** get the size of memory required to retrieve a string from a data buffer this includes space for a null terminator*/
+HELICS_EXPORT int helicsDataBufferStringSize(HelicsDataBuffer data);
+
+/* convert a data buffer to a string including a null terminator
+NOTE:  data may contain 0 prior to the end but actualLength will the number of characters in the string
```
Comment (nightlark) [inline src/helics/shared_api_library/helicsData.h:102]:
```suggestion
NOTE:  data may contain null bytes prior to the end but actualLength will be the number of characters in the string
```

**Review by nightlark — APPROVED** [review]

**Pull request merged**